src/neo4j_genai/kg_construction/schema.py

```diff
@@ -0,0 +1,99 @@
+#  Copyright (c) "Neo4j"
+#  Neo4j Sweden AB [https://neo4j.com]
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#  Copyright (c) "Neo4j"
+#  Neo4j Sweden AB [https://neo4j.com]
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from __future__ import annotations
+
+from typing import Any
+
+from neo4j_genai.exceptions import SchemaValidationError
+from pydantic import BaseModel, Field, model_validator
+
+
+class Entity(BaseModel):
+    name: str
+    type: str = Field(
+        ..., description="Type of the entity's name field, represented as a string."
+    )
+    description: str
+
+
+class Relation(BaseModel):
+    name: str
+    description: str
+
+
+class SchemaConfig(BaseModel):
+    entities: dict[str, dict[str, Any]]
+    relations: dict[str, dict[str, Any]]
+    potential_schema: dict[str, list[str]]
+
+    @model_validator(mode="before")
+    def check_schema(cls, data: dict[str, Any]) -> dict[str, Any]:
+        entities = data.get("entities", {}).keys()
+        relations = data.get("relations", {}).keys()
+        potential_schema = data.get("potential_schema", {})
+
+        for entity in potential_schema.keys():
+            if entity not in entities:
+                raise SchemaValidationError(
+                    f"Entity '{entity}' is not defined in the provided entities."
+                )
+
+        for rels in potential_schema.values():
+            for rel in rels:
+                if rel not in relations:
+                    raise SchemaValidationError(
+                        f"Relation '{rel}' is not defined in the provided relations."
+                    )
+
+        return data
+
+
+class SchemaBuilder:
```

Comment:
Could it be work implementing some sort of __str__ function here? I'm sure BaseModel has one but maybe we want to experiment with the string format of the schema somewhat when we're engineering the prompt for the entity extractor.

Reply:
Interesting, how does that work in the overall experience. In `kg_builder_with_schema.py` the schema inputs are defined before `asyncio.run(pipe.run(pipe_inputs))` is run. Won't a `__str__` only be useful if it can run when defining the pipeline. Did I misunderstand it?

Reply:
It could be useful I guess to have something similar in the `SchemaConfig` object, returned by the schema builder component. But for now the returned types are lost (we're just saving the dict), so at the moment we can't use it to serialize the schema for the prompt.

Reply:
Maybe this is something we want to do in the ER extractor instead in that case?

Reply:
Yes, to me that belongs to the prompt "builder" phase: you input the schema as a dict, and in the prompt you format it as you want (hence the idea of jinja2).